site/bps/BP-35-128-bits-support.md

```diff
@@ -0,0 +1,399 @@
+
+Currently DBLedgerStorage use `<ledgerId, entryId>` as the index key for indexing entry
+locations for each entry.
+
+Similar as `SortedLedgerStorage` and `InterleavedLedgerStorage`, for ledgers whose
+`ledgerScopeId` is 0, they will be using existing scheme for storing their entry locations.
+
+For ledgers whose `ledgerScopeId` is not 0, they will be stored in a new rocksdb,
+whose index key will be `<ledgerScopeId, ledgerId, entryId>`.
+
+#### Metadata Store
+
+##### LedgerManager
+
+All the interfaces should be updated with accepting `ledgerScopeId`.
+
+The actual implementation should decide how to store metadata
+for `<ledgerScopeId, ledgerId>`. 
+
+###### ZooKeeper Ledger Manager
+
+We need to introduce a LongLongHierchicalLedgerManager for storing metadata
+indexing by `<ledgerScopeId, ledgerId>`.
+
+If `ledgerScopeId` is 0, then it will be falling back to `LongHierachicalLedgerManager`.
+So no behavior is changed.
+
+If `ledgerScopeId` is not 0, those ledgers will be indexed in new hierarchy
+(possible under a different znode).
+
+###### Ledger ID generation
+
+When upgrading from 64bit to 128bits, we probably don't need any centralized mechanism
+for generating ledger id. It can be implemented using UUID generation.
+
+Especially since we are supporting 128bits by introducing `ledgerScopeId`. That means
+application of bookkeeper can decide its own way for generating their `scopeId`.
+An application or even bookkeeper client can generate its ledgerId using UUID generation,
+then breaks the 128 bits UUID into two parts, one serves as `ledgerScopeId` and the other
+one serves as `ledgerId`.
+
+###### Etcd
+
+Since Etcd has a better key/value presentation, we can basically just combine
+`<ledgerScopeId, ledgerId>` as the index key for storing ledger metadata in Etcd.
+Nothing is needed for special consideration.
+
+### Performance Concerns
+
+There shouldn't be any performance difference when not using 128 bit ledger id
+(`ledgerScopeId` is omitted).
+
+Performance concerns can be arised in following areas:
+
+- **Wire Protocol**: additional 9 bytes will be added per entry, one byte for version
+  and 8 bytes for the msb of 128 bit ledger id
+- **Journal**: additional 9 bytes will be added per entry (same as wire protocol). 
+- **EntryLogger**: additional 9 bytes will be added per entry (same as wire protocol)
+- **Memtable**: additional 8 bytes will be added per indexed entry.
+- **FileInfo**: there is no change to the index file format itself.
+- **IndexPersistenceManager**: Files will be organized in more directory hierarchy.
+  It shouldn't be a big deal. 
+- **IndexInMemoryManager (LedgerCache)**: additional 8 bytes per index page.
+- **DbLedgerStorage**: additional 8 bytes per entry for entry location.
+- **Metadata**: on zookeeper, we need a 128 bit ledger manager, that means more znode
+  hierarchy than 64 bit ledger manager. Etcd like key/value metadata store is probably
+  more preferrable for 128 bit ledger manager.
+
```

Comment:
Garbage collection?

Reply:
GC should remain unchanged, right? All the performance impact should be on metadata side or bookie storage side.

Reply:
From a performance PoV yes, but it will still need to take the scopeId into account.